File: code_samples/section5/problems/problem5_5/problem5_5.py
"""Problem 5.5: Lowest Common Ancestor"""

from typing import Optional

# pylint: disable=too-few-public-methods


class TreeNode:
    """
    Model for a binary tree node.

    Each node stores:
    - val   : the integer value at this node
    - left  : reference to the left child (or None)
    - right : reference to the right child (or None)
    """
    def __init__(
        self,
        val: int = 0,
        left: "Optional[TreeNode]" = None,
        right: "Optional[TreeNode]" = None
    ):
        self.val = val
        self.left = left
        self.right = right


def lowest_common_ancestor(
    root_node: Optional[TreeNode],
    p: TreeNode,
    q: TreeNode
) -> Optional[TreeNode]:
    """
    Return the Lowest Common Ancestor (LCA) of p and q in a BST.

    IMPORTANT: This solution assumes the tree is a *Binary Search Tree* (BST).

    BST property:
    - All values in left subtree  < node.val
    - All values in right subtree > node.val

    Key idea:
    Starting at the root, compare p and q to the current node:
    - If BOTH p and q are smaller than node.val, LCA must be in the left subtree.
    - If BOTH p and q are larger than node.val, LCA must be in the right subtree.
    - Otherwise (they split, or one equals the node), the current node is the LCA.
      This is the “first split point” on the path from root downward.

    Runs in O(h) time, where h is tree height, and O(1) extra space.
    """
    node = root_node

    # Cache values to avoid repeatedly reading p.val / q.val in the loop
    pv = p.val
    qv = q.val

    # Walk down the BST until we find the split point
    while node is not None:
        # If both targets are smaller, LCA must be to the left
        if pv < node.val and qv < node.val:
            node = node.left

        # If both targets are larger, LCA must be to the right
        elif pv > node.val and qv > node.val:
            node = node.right

        # Otherwise, node is the split point (or equals one target) => LCA found
        else:
            return node

    # If we somehow fall off the tree, return None (e.g., root_node was None)
    return None


# ----------------------------
# Tree printer (sideways)
# ----------------------------

def print_tree(root_node: Optional[TreeNode]):
    """
    Print the tree sideways for visualization.

    Convention:
    - Right subtree prints first (appears "up")
    - Left subtree prints last (appears "down")
    """
    print("Tree (sideways, right is up):")
    _print_tree_impl(root_node, 0)
    print()


INDENT = 4  # spaces per depth level when printing


def _print_tree_impl(node: Optional[TreeNode], indent: int):
    """Recursive helper for sideways tree printing."""
    if node is None:
        return

    _print_tree_impl(node.right, indent + INDENT)

    print(" " * indent + str(node.val))

    _print_tree_impl(node.left, indent + INDENT)


# ----------------------------
# Test utility
# ----------------------------

def test_lca(label: str, root_node: Optional[TreeNode], p: TreeNode, q: TreeNode):
    """
    Print a labeled LCA test:
    - shows which two nodes we're querying
    - prints the LCA value (or None)
    """
    print(f"==== {label} ====")
    ans = lowest_common_ancestor(root_node, p, q)
    print(f"LCA({p.val}, {q.val}) ->", ans.val if ans else None)
    print()


# ----------------------------
# Build BST for testing
# ----------------------------
#
#            6
#          /   \
#         2     8
#        / \   / \
#       0   4 7   9
#          / \
#         3   5
#
# This is a BST, so we can use the fast iterative split-point approach.
root = TreeNode(6)
root.left = TreeNode(2)
root.right = TreeNode(8)

root.left.left = TreeNode(0)
root.left.right = TreeNode(4)
root.left.right.left = TreeNode(3)
root.left.right.right = TreeNode(5)

root.right.left = TreeNode(7)
root.right.right = TreeNode(9)

# References for convenience (these are the actual node objects in the tree)
n2 = root.left
n8 = root.right
n4 = root.left.right
n3 = n4.left
n5 = n4.right
n0 = root.left.left
n7 = root.right.left
n9 = root.right.right


# ----------------------------
# Run test cases
# ----------------------------

# Print tree visually so test output is easier to follow
print_tree(root)

test_lca("LCA(2, 8) → expected 6", root, n2, n8)
test_lca("LCA(2, 4) → expected 2", root, n2, n4)
test_lca("LCA(3, 5) → expected 4", root, n3, n5)
test_lca("LCA(0, 5) → expected 2", root, n0, n5)
test_lca("LCA(7, 9) → expected 8", root, n7, n9)

File: code_samples/section5/problems/problem5_5/problem5_5.java
package code_samples.section5.problems.problem5_5;

// ==========================
// BINARY SEARCH TREE NODE
// ==========================

// Represents a single node in a Binary Search Tree (BST)
class TreeNode {
    int val;        // value stored in the node
    TreeNode left;  // reference to left child
    TreeNode right; // reference to right child

    // Constructor initializes the node value
    TreeNode(int v) {
        val = v;
    }
}

public class problem5_5 {

    // ==========================
    // LOWEST COMMON ANCESTOR (BST)
    // ==========================

    // Returns the Lowest Common Ancestor (LCA) of nodes p and q in a BST.
    //
    // BST properties used:
    // - All values in the left subtree  < node.val
    // - All values in the right subtree > node.val
    //
    // Algorithm:
    // - Start at the root.
    // - If BOTH p and q are smaller than the current node,
    //   move to the left subtree.
    // - If BOTH p and q are larger than the current node,
    //   move to the right subtree.
    // - Otherwise, the current node is the split point and
    //   therefore the LCA.
    //
    // Time Complexity: O(h), where h is the height of the tree
    // Space Complexity: O(1), iterative approach
    TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode node = root;

        // Cache values for easier comparisons
        int pv = p.val;
        int qv = q.val;

        // Walk down the tree until the LCA is found
        while (node != null) {

            // Both targets are in the left subtree
            if (pv < node.val && qv < node.val) {
                node = node.left;

            // Both targets are in the right subtree
            } else if (pv > node.val && qv > node.val) {
                node = node.right;

            // Split point found (or node equals p or q)
            } else {
                return node;
            }
        }

        // If the tree is empty or nodes are not present
        return null;
    }

    /* -------------------------------------------------------
       Helpers
       ------------------------------------------------------- */

    // Convenience method to create a new tree node
    private TreeNode node(int v) {
        return new TreeNode(v);
    }

    // Number of spaces to indent per tree level
    private static final int INDENT = 4;

    // Prints the tree sideways for visualization
    // Right subtree appears "up", left subtree appears "down"
    private void printTree(TreeNode root) {
        System.out.println("Tree (sideways, right is up):");
        printTreeImpl(root, 0);
        System.out.println();
    }

    // Recursive helper for sideways tree printing
    private void printTreeImpl(TreeNode root, int indent) {
        if (root == null) return;

        // Print right subtree first
        printTreeImpl(root.right, indent + INDENT);

        // Print current node with indentation
        System.out.println(" ".repeat(indent) + root.val);

        // Print left subtree
        printTreeImpl(root.left, indent + INDENT);
    }

    // Runs and prints a single LCA test case
    private void testLCA(String label, TreeNode root, TreeNode p, TreeNode q) {
        System.out.println("==== " + label + " ====");
        System.out.println("Find LCA(" + p.val + ", " + q.val + ")");

        TreeNode ans = lowestCommonAncestor(root, p, q);

        if (ans == null)
            System.out.println("LCA = null\n");
        else
            System.out.println("LCA = " + ans.val + "\n");
    }

    /* -------------------------------------------------------
       Main Test Harness
       ------------------------------------------------------- */

    public static void main(String[] args) {
        problem5_5 p = new problem5_5();

        /*
               6
             /   \
            2     8
           / \   / \
          0   4 7   9
             / \
            3   5
        */

        // Build the BST
        TreeNode root = p.node(6);
        root.left = p.node(2);
        root.right = p.node(8);

        root.left.left = p.node(0);
        root.left.right = p.node(4);
        root.left.right.left = p.node(3);
        root.left.right.right = p.node(5);

        root.right.left = p.node(7);
        root.right.right = p.node(9);

        // Visualize the tree
        p.printTree(root);

        // References to specific nodes in the tree
        TreeNode n2 = root.left;
        TreeNode n8 = root.right;
        TreeNode n4 = root.left.right;
        TreeNode n3 = root.left.right.left;
        TreeNode n5 = root.left.right.right;
        TreeNode n0 = root.left.left;
        TreeNode n7 = root.right.left;
        TreeNode n9 = root.right.right;

        // Run LCA test cases
        p.testLCA("LCA(2, 8) → expected 6", root, n2, n8);
        p.testLCA("LCA(2, 4) → expected 2", root, n2, n4);
        p.testLCA("LCA(3, 5) → expected 4", root, n3, n5);
        p.testLCA("LCA(0, 5) → expected 2", root, n0, n5);
        p.testLCA("LCA(7, 9) → expected 8", root, n7, n9);
    }
}
